### PR TITLE
Ref/mettalog

### PR DESCRIPTION
**Pull request by Biruk-gebru**
## Title

Speed up instance and list sorting

## Description

- Refactored sorting of instances, instance sets, and scored instances to use a fast `selectionSort` algorithm, operating on their expression forms for efficiency.
- Updated all relevant list methods in `list-methods.metta` to convert lists to expressions before sorting, and back to lists after sorting, leveraging `selectionSort` for speed.
- Added and updated tests in `utilities/tests/list-methods-test.metta` to cover the new sorting and conversion logic, including `List.sort`, `List.partialSort`, and expression conversion functions.
- Introduced and used a custom subtraction atom/function to replace the default subtraction, which behaved incorrectly in MetaLog.
- Improved composite score comparison by updating `cScoreExpr<` to compare score expressions directly, using `index-atom` to extract penalized and complexity scores from expressions.
- Added comprehensive tests for `cScoreExpr<` in `scoring/test/cscore-test.metta`, including round-trip conversion and edge cases (e.g., NaN handling, tie-breaking).

## Motivation and Context

- The primary motivation was to significantly improve the speed of sorting operations on lists, instances, and related data structures.
- The default subtraction atom in MetaLog was producing unexpected results, so a custom subtraction function was implemented and used throughout the codebase.
- The `cScoreExpr<` comparator now works directly on expressions, making score-based sorting and comparison more efficient and robust.
- These changes ensure both correctness and performance, especially in population management, candidate selection, and scoring routines.

## How Has This Been Tested?

- All changes are covered by updated and new tests in:
  - `metta-moses/representation/tests/instance.metta`
  - `metta-moses/utilities/tests/list-methods.metta`
  - `metta-moses/scoring/test/cscore.metta`
  - `metta-moses/representation/tests/instance-test.metta`
  - `metta-moses/utilities/tests/list-methods-test.metta`
  - `metta-moses/scoring/test/cscore-test.metta`
- Tests include round-trip conversions, sorting correctness, edge cases for the new subtraction atom, and direct expression-based score comparisons.
- Verified that all tests pass and that sorting and comparison are now faster and more reliable.

## Types of changes

- [x] Performance improvement (non-breaking change which improves speed)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: representation/add-logical-knobs.metta.pl
Comment (eyuuab) [inline representation/add-logical-knobs.metta.pl:1]: Hey @Biruk-gebru, before making any changes, please pull the latest updates from the ref/mettalog branch.
The .ignore file has been updated to exclude .pl and .html files.

File: representation/tests/instance-testold.metta
Comment (eyuuab) [inline representation/tests/instance-testold.metta:1]: Also, once you've made your changes to the test files, please rename them from `*testold.metta` to `*test.metta ` so they get picked up by the workflow for testing.

File: utilities/tests/list-methods-test.metta [hunk 1]
```diff
@@ -43,7 +43,14 @@
 !(assertEqual (List.sort (Cons 3 (Cons 2 (Cons 3 Nil)))) (Cons 2 (Cons 3 (Cons 3 Nil))))
 !(assertEqual (List.sort (Cons 5 (Cons 3 (Cons 1 (Cons 4 Nil))))) (Cons 1 (Cons 3 (Cons 4 (Cons 5 Nil)))))
 !(assertEqual (List.sort (Cons 9 (Cons 7 (Cons 8 (Cons 6 Nil))))) (Cons 6 (Cons 7 (Cons 8 (Cons 9 Nil)))))
-
+; !(assertEqual
```
[the remark this answers is not in the document]
Comment (Biruk-gebru) [reply utilities/tests/list-methods-test.metta]: This was a test case I had used to test the sorting on large values, but I've removed it now since it won't be necessary.

File: scoring/test/cscore-test.metta [hunk 2]
```diff
@@ -37,10 +43,76 @@
 ! (assertEqual (cScore< (mkCscore -0.1 2 0.1 0.1 -0.3) (mkCscore -0.2 3 0.1 0.1 -0.4)) False)
 ! (assertEqual (cScore< (mkCscore -0.5 3 0.1 0.1 -0.7) (mkCscore -0.5 2 0.1 0.1 -0.7)) True)
 ! (assertEqual (cScore< (mkCscore -0.5 2 0.1 0.1 -0.7) (mkCscore -0.5 3 0.1 0.1 -0.7)) False)
-! (assertEqual (cScore< (mkCscore -0.5 2 0.1 0.1 1.5NaN) (mkCscore -0.5 3 0.1 0.1 -0.7)) True)
+; ! (assertEqual (cScore< (mkCscore -0.5 2 0.1 0.1 nan) (mkCscore -0.5 3 0.1 0.1 -0.7)) True)
```
[the remark this answers is not in the document]
Comment (Biruk-gebru) [reply scoring/test/cscore-test.metta:46]: These were tests I was having issues with and had left commented. They have now been uncommented and restored.
